language/libra-vm/src/on_chain_configs.rs

```diff
@@ -0,0 +1,67 @@
+use libra_types::{on_chain_config::access_path_for_config, transaction::SCRIPT_HASH_LENGTH};
+use move_core_types::identifier::Identifier;
+use move_vm_state::data_cache::RemoteCache;
+use once_cell::sync::Lazy;
+use serde::{Deserialize, Serialize};
+use std::collections::BTreeSet;
+
+#[derive(Debug, Clone)]
+pub(crate) struct VMConfig {
+    pub publishing_options: VMPublishingOption,
+}
+
+/// Defines and holds the publishing policies for the VM. There are three possible configurations:
+/// 1. No module publishing, only whitelisted scripts are allowed.
+/// 2. No module publishing, custom scripts are allowed.
+/// 3. Both module publishing and custom scripts are allowed.
+/// We represent these as an enum instead of a struct since whitelisting and module/script
+/// publishing are mutually exclusive options.
+#[derive(Clone, Debug, Deserialize, PartialEq, Serialize)]
+pub enum VMPublishingOption {
+    /// Only allow scripts on a whitelist to be run
+    Locked(BTreeSet<[u8; SCRIPT_HASH_LENGTH]>),
+    /// Allow custom scripts, but _not_ custom module publishing
+    CustomScripts,
+    /// Allow both custom scripts and custom module publishing
+    Open,
+}
+
+impl VMConfig {
+    pub fn load_on_chain_config(state_view: &dyn RemoteCache) -> Option<Self> {
```

Comment:
it really feels like these should be errors and not options. shouldn't it be an error not to have an onchain config if it were expected? It could be due to bad serialization or it is missing. An option just says there isn't one.

Reply:
I can imagine when execute genesis txn, the on-chain-config is not in state view yet

Reply:
Sure, but couldn't that be executed on a different code path, one that is explicitly intended for genesis to setup a non-existent config? That way you eliminate all the conditional checks. In steady state, it either exists or you fail.